src/ray/core_worker/experimental_mutable_object_manager.cc

```diff
@@ -14,29 +14,97 @@
 
 #include "ray/core_worker/experimental_mutable_object_manager.h"
 
+// On Darwin, named semaphores cannot have names longer than PSEMNAMLEN. On other
+// platforms, we define PSEMNAMLEN to be 31, which is its current value in Darwin.
+#ifdef __APPLE__
+#include <sys/posix_sem.h>
+#elif !defined(PSEMNAMLEN)
+#define PSEMNAMLEN 31UL
+#endif
+
+#include "absl/strings/str_format.h"
+
 namespace ray {
 
+ExperimentalMutableObjectManager::~ExperimentalMutableObjectManager() {
+  // Copy `semaphores_` into `tmp` because `CloseSemaphore()` mutates `semaphores_`.
+  absl::flat_hash_map<ObjectID, sem_t *> tmp = semaphores_;
+  for (const auto &[object_id, _] : tmp) {
+    CloseSemaphore(object_id);
+  }
+}
+
 Status ExperimentalMutableObjectManager::RegisterWriterChannel(
     const ObjectID &object_id, std::unique_ptr<plasma::MutableObject> mutable_object) {
-#ifdef __linux__
   auto inserted =
       writer_channels_.emplace(object_id, WriterChannel(std::move(mutable_object)));
   if (!inserted.second) {
     return Status::Invalid("Writer channel already registered");
   }
   RAY_CHECK(inserted.first->second.mutable_object);
+  OpenSemaphore(object_id);
   RAY_LOG(DEBUG) << "Registered writer channel " << object_id;
-#endif
   return Status::OK();
 }
 
+std::string ExperimentalMutableObjectManager::GetSemaphoreName(
+    const ObjectID &object_id) {
+  PlasmaObjectHeader *header = nullptr;
+  {
+    auto it = writer_channels_.find(object_id);
+    if (it != writer_channels_.end()) {
+      header = it->second.mutable_object->header;
+    }
+  }
+  {
+    auto it = reader_channels_.find(object_id);
+    if (it != reader_channels_.end()) {
+      header = it->second.mutable_object->header;
+    }
+  }
+  RAY_CHECK(header);
+
+  static const char zero[PSEMNAMLEN] = {0};
+  if (!memcmp(header->unique_name, zero, sizeof(zero))) {
```

Comment:
I think it would be better to put this in the Init() code for the PlasmaObjectHeader. When we do Init(), we're guaranteed that there's only one writer (the plasma store process), whereas here there may be multiple writers.